Sources/Core/DependencyFinder.swift

```diff
@@ -0,0 +1,101 @@
+
+import Foundation
+
+final class DependencyFinder: DependencyFinding {
+
+    enum DependencyFinderError: Error, LocalizedError {
+        case failedCollectingDependencies(packageUrl: URL)
+        case missingVersionRef(dependency: String)
+
+        var errorDescription: String? {
+            switch self {
+            case .failedCollectingDependencies(let packageUrl):
+                return "Failed collecting dependencies at \(packageUrl.path)"
+            case .missingVersionRef(let dependency):
+                return "Missing versionRef for \(dependency)"
+            }
+        }
+    }
+
+    private let fileManager: FileManager
+
+    // MARK: - Inits
+
+    init(fileManager: FileManager = .default) {
+        self.fileManager = fileManager
+    }
+
+    // MARK: - Functions
+
+    func findPackageDependencies(at url: URL, versionRefsPath: String) async throws -> [PackageDependency] {
+        let process = Process()
+        process.executableURL = URL(filePath: "/usr/bin/swift", directoryHint: .notDirectory)
+        process.arguments = ["package", "show-dependencies", "--format", "json"]
+        process.currentDirectoryURL = url.absoluteURL
+
+        let pipe = Pipe()
+        process.standardOutput = pipe
+        process.standardError = Pipe()
+
+        try process.run()
+        process.waitUntilExit()
+
+        guard process.terminationStatus == 0 else {
+            throw DependencyFinderError.failedCollectingDependencies(packageUrl: url)
+        }
+
+        let data = pipe.fileHandleForReading.readDataToEndOfFile()
+        let packageDescription = try JSONDecoder().decode(PackageSpec.self, from: data)
+        return try parseDependencies(
+            packageDescription.dependencies,
+            versionRefsPath: versionRefsPath
+        )
+    }
+
+    // MARK: - Helper Functions
+
+    private func parseDependencies(_ dependencies: [DependencySpec], versionRefsPath: String) throws -> [PackageDependency] {
+        var result = [PackageDependency]()
+        let dependencyReferences = try loadDependencyReferences(from: versionRefsPath)
+
+        for dependency in dependencies {
+            if dependency.version == "unspecified" {
+                guard let versionRef = dependencyReferences.first(where: { dep in
+                    dep.name == dependency.name
+                })?.versionRef else {
+                    throw DependencyFinderError.missingVersionRef(dependency: dependency.name)
+                }
+                result.append(
+                    PackageDependency(
+                        name: dependency.name,
+                        type: .local(hash: versionRef)
+                    )
+                )
+            }
+            else {
+                result.append(
+                    PackageDependency(
+                        name: dependency.name,
+                        type: .remote(tag: dependency.version)
+                    )
+                )
+            }
+
```

Comment:
both if and else has similar code 
```
result.append(
                    PackageDependency(
                        name: dependency.name,
                        type: .remote(tag: dependency.version)
                    )
                )
```
I would suggest to simplify:
```
let type = if ... {
  .local(hash: versionRef)
} else {
  .remote(tag: dependency.version)
}

result.append(
  PackageDependency(
    name: dependency.name,
    type: type
  )
)
```

Reply:
There is a guard in the way. This is meant to stay like this for readability.

Reply:
Pushed ac64d1de5da35a6c570aa8bf27545d254bb5476c